### PR TITLE
[Fix] eagle/eagle3 speculative decoding conflicts with xgrammar in NPU

### PR DESCRIPTION
**Pull request by ChefWu551**
## Motivation

Fix when turn on the eagle/eagle3 spec-decoding and send a request with json_schema, as reported in [#20148](https://github.com/sgl-project/sglang/issues/20148).
The reason is that , on Ascend NPU, apply_token_bitmask_inplace_triton can fail to enforce grammar masks for XGrammar-constrained decoding: tokens that are masked out (mask_allows=False) may still keep high logits and be selected, causing downstream Tokens not accepted errors in grammar.accept_token.

This PR supersedes a previously closed PR (#20168),
which was accidentally invalidated due to a force push.

This version reintroduces the changes on a clean branch.

## Modifications

This PR adds an NPU-safe fallback for constrained mask application:

For logits.device.type == "npu" in XGrammar, use a CPU-based bitmask unpack + NPU masked_fill(-inf) path.
Keep existing Triton/CUDA paths unchanged for non-NPU devices.
Move the NPU fallback helper into triton_ops/bitmask_ops.py for centralized mask-op management.

## Accuracy Tests
server responses correctly and the response keeps consistent with the case where eagle is turned off.

02d99b3d08ee824b9a451f032314c196

## Benchmarking and Profiling
1d8fb94bc89737a4df7b047789223973

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/constrained/triton_ops/bitmask_ops.py [hunk 1]
```diff
@@ -139,3 +139,21 @@ def apply_token_bitmask_inplace_triton(
...
         num_stages=3,
     )
+
+
+def apply_token_bitmask_inplace_torch_npu(
+    logits: torch.Tensor,
+    bitmask: torch.Tensor,
+):
+    """NPU-safe fallback for packed-bitmask application."""
+    assert logits.device.type == "npu", "NPU-only helper called on non-NPU device"
```
Comment (kpham-sgl) [inline python/sglang/srt/constrained/triton_ops/bitmask_ops.py]: I don't think theres any logic in this function specific to NPU (and neither it is triton). It works as a normal CPU + PyTorch fallback option for any backend.  

I propose
- Rename and move it away from `triton_ops/`
- Remove the assertion and left the comment saying it is being used as a fallback for NPU

**Review by kpham-sgl — CHANGES_REQUESTED** [review]
@ChefWu551 agree on the problem and general solution. Left one comment.
There's also a recent merge in `xgrammar_backend.py` to support MUSA backend so there is a minor conflict. 
Can you address the comment and resolve the conflicts? Thanks!

File: python/sglang/srt/constrained/triton_ops/bitmask_ops.py (hunk 1 above)
Comment (ChefWu551) [reply python/sglang/srt/constrained/triton_ops/bitmask_ops.py]: Thanks for the review and suggestion. I’ve updated the PR accordingly:

- moved the torch fallback out of triton_ops/ to constrained/torch_ops/bitmask_ops.py
- renamed it to a backend-agnostic fallback (apply_token_bitmask_inplace_torch)
- removed the NPU-only assertion
- kept a note that it is currently used as the NPU fallback in xgrammar_backend.py

**Comment by ChefWu551** [comment]
> @ChefWu551 agree on the problem and general solution. Left one comment. There's also a recent merge in `xgrammar_backend.py` to support MUSA backend so there is a minor conflict. Can you address the comment and resolve the conflicts? Thanks!

I've resolved the conflict issue. Could you please review it again for me?

**Review by kpham-sgl — APPROVED** [review]
LGTM

**Review by hnyls2002 — APPROVED** [review]
